File: 8_kyu/Total_amount_of_points.c
#include <string.h>

const int NUMBER_GAMES = 10;

// Sum the points reached in the NUMBER_GAMES games.
// Each game has format "x:y".
int points(const char* const games[NUMBER_GAMES]) {
  // NULL pointer error.
  if( games == NULL )
    return 0;
  int sum = 0;
  for( int i = 0; i < NUMBER_GAMES; ++i ) {
    // Format error.
    if( games[i] == NULL || strlen(games[i]) != 3 || games[i][1] != ':' )
      return 0;
    // Add points for the game i.
    const char x = games[i][0];
    const char y = games[i][2];
    if( x > y )
      sum += 3;
    else if( x == y )
      sum += 1;
  }
  return sum;
}

File: 8_kyu/Total_amount_of_points.rs
use std::error::Error;

// Calculates the total amount of points.
// Each string must be in format "a:b", invalid formated strings are ignored.
// Point system is win: 3, tie: 1, loss: 0.
fn points(games: &[String]) -> u32 {
    games
        .iter()
        .map(gameresult_from_str)
        .map(|result| match result {
            Ok(result) if result.0 > result.1 => 3,
            Ok(result) if result.0 == result.1 => 1,
            _ => 0,
        })
        .sum()
}

// Stores the result of a game like the "a:b" string.
struct GameResult(u32, u32);

// Converts a game result like "a:b" with to decimal numbers a,b int a GameResult.
fn gameresult_from_str(str: &String) -> Result<GameResult, Box<dyn Error>> {
    let parts: Vec<&str> = str.split(':').collect();
    if parts.len() != 2 {
        return Err("not a:b format".into());
    }
    Ok(GameResult(
        parts[0].parse::<u32>()?,
        parts[1].parse::<u32>()?,
    ))
}
